File: articles/monte-carlo-methods-and-computational-uncertainty/calculators/python/model_calculator.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import math
import random
from statistics import mean, pstdev


def ci(values: list[float]) -> tuple[float, float, float, float]:
    n = len(values)
    estimate = mean(values)
    sd = pstdev(values) if n > 1 else 0.0
    se = sd / math.sqrt(n) if n > 1 else 0.0
    return estimate, se, estimate - 1.96 * se, estimate + 1.96 * se


def pi_estimate(samples: int, seed: int) -> None:
    rng = random.Random(seed)
    vals = []
    for _ in range(samples):
        x = rng.random(); y = rng.random()
        vals.append(4.0 if x * x + y * y <= 1.0 else 0.0)
    estimate, se, lower, upper = ci(vals)
    print(f"pi_estimate={estimate:.8f}")
    print(f"standard_error={se:.8f}")
    print(f"lower_95={lower:.8f}")
    print(f"upper_95={upper:.8f}")
    print(f"absolute_error={abs(estimate - math.pi):.8f}")


def threshold_risk(samples: int, seed: int, threshold: float) -> None:
    rng = random.Random(seed)
    vals = []
    for _ in range(samples):
        base = rng.triangular(900000, 1100000, 1000000)
        labor = max(0.75, rng.gauss(1.0, 0.08))
        delay = max(0.0, rng.gauss(60000, 35000))
        contingency = rng.uniform(20000, 120000)
        vals.append(1.0 if base * labor + delay + contingency > threshold else 0.0)
    estimate, se, lower, upper = ci(vals)
    print(f"threshold_probability={estimate:.6f}")
    print(f"standard_error={se:.6f}")
    print(f"lower_95={lower:.6f}")
    print(f"upper_95={upper:.6f}")


def main() -> None:
    parser = argparse.ArgumentParser(description="Monte Carlo uncertainty calculator")
    parser.add_argument("mode", choices=["pi", "threshold"])
    parser.add_argument("--samples", type=int, default=10000)
    parser.add_argument("--seed", type=int, default=42)
    parser.add_argument("--threshold", type=float, default=1250000)
    args = parser.parse_args()
    if args.mode == "pi":
        pi_estimate(args.samples, args.seed)
    else:
        threshold_risk(args.samples, args.seed, args.threshold)


if __name__ == "__main__":
    main()

File: articles/monte-carlo-methods-and-computational-uncertainty/calculators/r/model_calculator.R
#!/usr/bin/env Rscript
args <- commandArgs(trailingOnly = TRUE)
mode <- ifelse(length(args) >= 1, args[1], "pi")
samples <- 10000
seed <- 42
threshold <- 1250000

for (arg in args) {
  if (grepl("^--samples=", arg)) samples <- as.integer(sub("^--samples=", "", arg))
  if (grepl("^--seed=", arg)) seed <- as.integer(sub("^--seed=", "", arg))
  if (grepl("^--threshold=", arg)) threshold <- as.numeric(sub("^--threshold=", "", arg))
}

ci <- function(values) {
  estimate <- mean(values)
  se <- if (length(values) > 1) sd(values) / sqrt(length(values)) else 0
  c(estimate = estimate, standard_error = se, lower_95 = estimate - 1.96 * se, upper_95 = estimate + 1.96 * se)
}

set.seed(seed)

if (mode == "pi") {
  x <- runif(samples)
  y <- runif(samples)
  values <- ifelse(x * x + y * y <= 1, 4, 0)
  out <- ci(values)
  cat(sprintf("pi_estimate=%.8f\n", out["estimate"]))
  cat(sprintf("standard_error=%.8f\n", out["standard_error"]))
  cat(sprintf("lower_95=%.8f\n", out["lower_95"]))
  cat(sprintf("upper_95=%.8f\n", out["upper_95"]))
  cat(sprintf("absolute_error=%.8f\n", abs(out["estimate"] - pi)))
} else if (mode == "threshold") {
  base <- runif(samples, 900000, 1100000)
  labor <- pmax(0.75, rnorm(samples, 1.0, 0.08))
  delay <- pmax(0, rnorm(samples, 60000, 35000))
  contingency <- runif(samples, 20000, 120000)
  values <- ifelse(base * labor + delay + contingency > threshold, 1, 0)
  out <- ci(values)
  cat(sprintf("threshold_probability=%.6f\n", out["estimate"]))
  cat(sprintf("standard_error=%.6f\n", out["standard_error"]))
  cat(sprintf("lower_95=%.6f\n", out["lower_95"]))
  cat(sprintf("upper_95=%.6f\n", out["upper_95"]))
} else {
  stop("Mode must be pi or threshold")
}
